File: loglib-java/src/main/java/com/robog/loglib/AbstractLog.java
package com.robog.loglib;

/**
 * Created by yuxingdong on 2018/6/22.
 */
abstract class AbstractLog implements LogStrategy {

    private final PutTask mPutTask;

    AbstractLog() {
        mPutTask = new PutTask();
    }

    @Override
    public void logAsync(LogBean logBean) {
        mPutTask.setLogBean(logBean);
        Schedulers.io(mPutTask);
    }

    @Override
    public void logSync(LogBean logBean) {
        LogCache.get().put(logBean);
    }

    static class PutTask implements Runnable {

        private LogBean mLogBean;

        void setLogBean(LogBean logBean) {
            mLogBean = logBean;
        }

        @Override
        public void run() {
            LogCache.get().put(mLogBean);
        }
    }
}

File: loglib-kotlin/src/main/java/com/robog/loglib/AbstractLog.kt
package com.robog.loglib

/**
 * Created by yuxingdong on 2018/6/29.
 */
internal abstract class AbstractLog : LogStrategy {

    private val putTask: PutTask

    init {
        putTask = PutTask()
    }

    override fun logAsync(logBean: LogBean) {
        putTask.logBean = logBean
        io {
            putTask.run()
        }
    }

    override fun logSync(logBean: LogBean) {
        LogCache.get().put(logBean)
    }

    internal class PutTask : Runnable {

        var logBean: LogBean? = null

        override fun run() {
            LogCache.get().put(checkNotNull(logBean))
        }
    }
}
